dev/src/path.ts

```diff
@@ -331,51 +383,52 @@ export class ResourcePath extends Path<ResourcePath> {
    * methods.
    *
    * @private
-   * @override
-   * @param {Array.<string>} segments Sequence of names of the parts of the
-   * path.
-   * @returns {ResourcePath} The newly created ResourcePath.
+   * @param segments Sequence of names of the parts of the path.
+   * @returns The newly created QualifiedResourcePath.
    */
-  construct(segments: string[]): ResourcePath {
-    return new ResourcePath(this.projectId, this.databaseId, ...segments);
+  construct(segments: string[]): QualifiedResourcePath {
+    return new QualifiedResourcePath(
+        this.projectId, this.databaseId, ...segments);
+  }
+
+  /**
+   * Convenience method to match the ResourcePath API. This method always
+   * returns the current instance. The arguments is ignored.
```

Comment:
I renamed the argument to `projectIdIfMissing` here and in ResourcePath. When I started writing this, I originally added an assert - but that doesn't work since the `projectId` in the argument is always the Project ID of the SDK's Firestore project. The Project ID in the path itself may be an ID from a different project if we deserialize a DocumentReference from a document.
An alternative would be to remove this method and only call `toQualifiedResourcePath` if we need to. That makes the callsites more complicated, and I kinda like the simplicity of the current approach. So for now I have just updated the argument name and comment, but please do let me know if you'd rather want me to drop this method altogether.

Reply:
Oh, wow.  So DocumentReference may wrap either a ResourcePath or a QualifiedResourcePath and so we have to make sure we preserve the projectId if it's already a QualifiedResourcePath, but insert the current projectId if it's not?  I didn't realize that issue (partially because we don't deal with it in the mobile SDK).  That is a bit funky, but I guess I don't feel strongly enough to suggest you rework this anymore...

Reply:
Another way to solve this would be to add Firestore as a member to ResourcePath. `formattedName()` could then figure out what it needs on its own. I'll leave this as an exercise for the reader for now.